compiler/src/iree/compiler/Dialect/Util/Transforms/ImportResources.cpp

```diff
@@ -0,0 +1,213 @@
+          // Already seen?
+          auto it = replacements.find(elements);
+          if (it != replacements.end()) {
+            LLVM_DEBUG(llvm::dbgs()
+                       << ":: Replacing already encountered attr of "
+                       << elements.getType() << "\n");
+            attr.setValue(it->second);
+            updated = true;
+            continue;
+          }
+
+          // Convert.
+          if (shouldConvertElements(elements)) {
+            LLVM_DEBUG(llvm::dbgs() << ":: Converting elements attr of "
+                                    << elements.getType() << "\n");
+            if (auto replacement = convertElementsAttr(elements)) {
+              attr.setValue(replacement);
+              replacements[elements] = replacement;
+              updated = true;
+            } else {
+              LLVM_DEBUG(llvm::dbgs() << "  Failed to convert\n");
+            }
+          }
+        }
+      }
+      if (updated)
+        op->setAttrs(attrs);
+    });
+    LLVM_DEBUG(llvm::dbgs() << "DONE CONVERTING RESOURCES\n");
+  }
+
+  static bool shouldConvertElements(ElementsAttr attr) {
+    if (llvm::isa<DenseElementsAttr>(attr)) {
+      // DenseElementsAttr encodes arbitrary dimension
+      // splats whereas DenseResourceElementsAttr does not.
+      if (attr.isSplat())
+        return false;
+      return true;
+    }
+
+    return false;
+  }
+
+  static ElementsAttr convertElementsAttr(ElementsAttr elementsAttr) {
+    auto st = llvm::cast<ShapedType>(elementsAttr.getType());
+    auto elementType = st.getElementType();
+    auto numElements = elementsAttr.getNumElements();
+    auto bitWidth = elementType.getIntOrFloatBitWidth();
+    AsmResourceBlob blob;
+    if (auto attr = llvm::dyn_cast<DenseIntElementsAttr>(elementsAttr)) {
+      switch (bitWidth) {
+      case 1:
+        blob = HeapAsmResourceBlob::allocate(numElements, /*align=*/64,
+                                             /*dataIsMutable=*/true);
+        copyIntAttrIntoBlob<uint8_t, /*numBits=*/1>(blob, attr);
+        return BlobResourceElementsAttr::get(st, "dense_elements_i1",
+                                             std::move(blob));
+      case 8:
+        blob = HeapAsmResourceBlob::allocate(numElements, /*align=*/64,
+                                             /*dataIsMutable=*/true);
+        copyIntAttrIntoBlob<uint8_t>(blob, attr);
+        return BlobResourceElementsAttr::get(st, "dense_elements_i8",
+                                             std::move(blob));
+      case 16:
+        blob = HeapAsmResourceBlob::allocate(2 * numElements, /*align=*/64,
+                                             /*dataIsMutable=*/true);
+        copyIntAttrIntoBlob<uint16_t>(blob, attr);
+        return BlobResourceElementsAttr::get(st, "dense_elements_i16",
+                                             std::move(blob));
+      case 32:
+        blob = HeapAsmResourceBlob::allocate(4 * numElements, /*align=*/64,
```

Comment:
Wouldn't the majority here be bitWidth/8? (Joining supported cases)

Reply:
Yeah, but there's some template goo in the copy that is needed to appease DenseElementsAttr...